prusti-viper/src/encoder/snapshot_encoder.rs

```diff
@@ -647,24 +737,694 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
+                    self.snapshot_encoder.encode_arg_local(SNAPSHOT_ARG),
+                    self.adt_def.non_enum_variant(),
+                    self.subst,
+                )?
+            )
+        } else {
+            let variant_arg = vir::Expr::field(
+                self.snapshot_encoder.encode_arg_local(SNAPSHOT_ARG),
+                self.snapshot_encoder.encoder.encode_discriminant_field(),
+            );
+
+            self.fold_snap_func_conditional(
+                snap_domain,
+                variant_arg,
+                0,
+            )?
+        };
+
+        Ok(
+            self.snapshot_encoder.encode_snap_func(
+                snap_domain.get_type(),
+                body,
+            )
+        )
+    }
+
+    fn fold_snap_func_conditional(
+        &self,
+        snap_domain: &SnapshotDomain,
+        variant_arg: vir::Expr,
+        index: usize,
+    ) -> EncodingResult<vir::Expr>
+    {
+        if index >= self.adt_def.variants.len() - 1 {
+            self.encode_snap_variant(snap_domain, index)
+        } else {
+            let tcx = self.snapshot_encoder.encoder.env().tcx();
+            let variant_index = self.adt_def.discriminant_for_variant(
+                tcx,
+                rustc_target::abi::VariantIdx::from_usize(index)
+            ).val;
+            let size = ty::tls::with(|tcx| Integer::from_attr(&tcx, self.adt_def.repr.discr_type()).size());
+            let discriminant = size.sign_extend(variant_index) as i128;
+
+            Ok(vir::Expr::ite(
+                vir::Expr::eq_cmp(
+                    variant_arg.clone(),
+                    vir::Expr::from(discriminant),
+                ),
+                self.encode_snap_variant(snap_domain, index)?,
+                self.fold_snap_func_conditional(
+                    snap_domain,
+                    variant_arg,
+                    index+1,
+                )?,
+            ))
+        }
+    }
+
+    fn encode_snap_variant(
+        &self,
+        snap_domain: &SnapshotDomain,
+        variant_index: usize,
+    ) -> EncodingResult<vir::Expr>
+    {
+        let (
+                variant_location,
+                predicate_name,
+                variant,
+                enum_variant_index
+        ) = self.obtain_variant(variant_index)?;
+
+        let cons_func = snap_domain.domain.functions[variant_index].clone();
+
+        if variant.fields.is_empty() {
+            Ok(vir::Expr::domain_func_app(cons_func, vec![]))
+        } else {
+            let args = self.snapshot_encoder.encode_cons_func_args(
+                variant_location.clone(),
+                variant,
+                self.subst
+            )?;
+            Ok(
+                vir::Expr::unfolding(
+                    predicate_name,
+                    vec![variant_location],
+                    vir::Expr::domain_func_app(cons_func, args),
+                    vir::PermAmount::Read,
+                    enum_variant_index,
+                )
```

Comment:
@tillarnold Why is this needed? Does the fold/unfold step fail to generate the variant unfolding automatically?

Reply:
I am not sure why this is needed. I think (if I did git blame correctly) this was originally added in commit https://github.com/viperproject/prusti-dev/commit/8e0bf653d7e114e25e7e7f36dbc16f9c497983b0 so I don't think I'm the original author of this.